### PR TITLE
The JavaScript Visitor now implements a depth-first traversal

### PR DESCRIPTION
**Pull request by cgudrian**
Just like the Java version.

### CODE REVIEW

**Comment by cgudrian** [comment]
cf. #1285

**Comment by ericvergnaud** [comment]
@cgudrian Hi, happy to bless this, if you can provide at least one test case, as described in #1275

**Comment by cgudrian** [comment]
I've extended the test case generator and implemented the JavaScript tests. The other languages are missing, though. Should I open a pull request nevertheless?

**Comment by ericvergnaud** [comment]
Hi,
there is an IgnoredTests section in each target language stg.
you may want to populate it for non javascript targets.
Eric

**Comment by ericvergnaud** [comment]
Hi,
it seems the JavaScript tests fail on Travis.
Eric

**Comment by cgudrian** [comment]
That looks better.

**Comment by cgudrian** [comment]
Is there anything left for me to do to have this request merged into master?

File: runtime-testsuite/test/org/antlr/v4/test/runtime/javascript/node/BaseTest.java [hunk 1]
```diff
@@ -413,7 +413,7 @@ public String execModule(String fileName) {
 	}
 
 	private String locateTool(String tool) {
-		String[] roots = { "/usr/bin/", "/usr/local/bin/" };
+		String[] roots = { "/usr/bin/", "/usr/local/bin/", "/Users/christian/.nvm/versions/node/v6.6.0/bin/" };
```
Comment (ericvergnaud) [inline runtime-testsuite/test/org/antlr/v4/test/runtime/javascript/node/BaseTest.java]: Not sure we need a reference to your particular nodes setup

File: tool/resources/org/antlr/v4/tool/templates/codegen/JavaScript/JavaScript.stg [hunk 2]
```diff
@@ -91,7 +91,6 @@ exports.<file.grammarName>Listener = <file.grammarName>Listener;
 VisitorFile(file, header) ::= <<
 <fileHeader(file.grammarFileName, file.ANTLRVersion)>
 var antlr4 = require('antlr4/index');
-<header>
```
Comment (ericvergnaud) [inline tool/resources/org/antlr/v4/tool/templates/codegen/JavaScript/JavaScript.stg]: Not sure why you removed header, can you elaborate?

Comment (cgudrian) [inline tool/resources/org/antlr/v4/tool/templates/codegen/JavaScript/JavaScript.stg]: The generated header looks like this:
```javascript
var TVisitor = require('./TVisitor').TVisitor;
```
which leads to `TVisitor` being `undefined` even after
```javascript
function TVisitor() {
    …
}
```
Consequently trying to access `TVisitor.prototype` yields an exception.

If you compare the VisitorFile template with the ListenerFile template you'll see, that the latter one lacks the `<header>` as well.

**Comment by cgudrian** [comment]
Should I rebase this branch to remove traces to my local node installation?

File: tool/resources/org/antlr/v4/tool/templates/codegen/JavaScript/JavaScript.stg (hunk 2 above)
Comment (ericvergnaud) [inline tool/resources/org/antlr/v4/tool/templates/codegen/JavaScript/JavaScript.stg]: Ok

**Comment by ericvergnaud** [comment]
I don't think it's required.

**Comment by ericvergnaud** [comment]
Blessed.

**Comment by ericvergnaud** [comment]
You need to add yourself to the contributors list

**Comment by cgudrian** [comment]
What's the preferred way of dealing with conflicts in pull requests: merge or rebase?

**Comment by parrt** [comment]
Heh guys. I'd like to keep all tests current for all targets. E.g., when I add a java test, i add it for the others as well.

**Pull request merged**